api/v1alpha4/openstackcluster_webhook.go

```diff
@@ -17,16 +17,64 @@ limitations under the License.
 package v1alpha4
 
 import (
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
 	"sigs.k8s.io/controller-runtime/pkg/builder"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
 )
 
 // log is for logging in this package.
 var _ = logf.Log.WithName("openstackcluster-resource")
 
+// OpenStackClusterSpecIdentityRefKindImmutableMsg is immutable.
+const OpenStackClusterSpecIdentityRefKindImmutableMsg = "OpenStackCluster spec.IdentityRef.kind field is immutable. Please create a new resource instead."
+
 func (r *OpenStackCluster) SetupWebhookWithManager(mgr manager.Manager) error {
 	return builder.WebhookManagedBy(mgr).
 		For(r).
 		Complete()
 }
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha4-openstackcluster,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=openstackcluster,versions=v1alpha4,name=validation.openstackcluster.infrastructure.x-k8s.io,sideEffects=None,admissionReviewVersions=v1beta1
+
+var _ webhook.Validator = &OpenStackCluster{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (r *OpenStackCluster) ValidateCreate() error {
+	var allErrs field.ErrorList
+
+	if r.Spec.IdentityRef != nil {
+		if r.Spec.IdentityRef.Kind != "Secret" {
+			allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "identityRef", "kind"), "must be a Secret"))
+		}
+	}
+
+	return aggregateObjErrors(r.GroupVersionKind().GroupKind(), r.Name, allErrs)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (r *OpenStackCluster) ValidateUpdate(old runtime.Object) error {
+	var allErrs field.ErrorList
+	oldOpenStackCluster := old.(*OpenStackCluster)
+
+	if !reflect.DeepEqual(r.Spec.IdentityRef.Kind, oldOpenStackCluster.Spec.IdentityRef.Kind) {
+		allErrs = append(allErrs,
+			field.Invalid(field.NewPath("spec", "identityRef", "kind"), r, OpenStackClusterSpecIdentityRefKindImmutableMsg),
+		)
+	}
+
+	if len(allErrs) != 0 {
+		return aggregateObjErrors(r.GroupVersionKind().GroupKind(), r.Name, allErrs)
+	}
+
+	return nil
```

Comment:
```suggestion
	return aggregateObjErrors(r.GroupVersionKind().GroupKind(), r.Name, allErrs)
```

I assume that should work (as the other webhooks are using it)

Reply:
Also just copied from `openstackmachinetemplate_webhook.go`.
Will fix it there, too